WhateverGreen/kern_igfx.cpp

```diff
@@ -651,6 +721,585 @@ int IGFX::wrapReadAUX(void *that, IORegistryEntry *framebuffer, uint32_t address
+	// and an additional API layer is implemented to provide R/W access to registers at specific offsets on the adapter.
+	// AppleIntelFramebufferController::GetDPCDInfo() is then wrapped to inject code for LSPCON driver initialization,
+	// so that the adapter is configured to run in PCON mode instead on plugging the HDMI 2.0 cable to the port.
+	// Besides, the adapter is able to handle HDMI 1.4 connections when running in PCON mode, so we don't need to switch
+	// back to LS mode again.
+	//
+	// If you are interested in the detailed theory behind this fix, please take a look at my blog post.
+	// [ToDo: The article is still working in progress. I will add the link at here once I finish it. :D]
+	//
+	// Notes:
+	// 1. This fix is applicable for all laptops and PCs with HDMI 2.0 routed to IGPU.
+	//    (Laptops/Mobiles start from Skylake platform. e.g. Intel Skull Canyon NUC; Iris Pro 580 and HDMI 2.0)
+	// 2. If your HDMI 2.0 with Intel (U)HD Graphics is working properly, you don't need this fix,
+	//    as the adapter might already be configured in the firmware to run in PCON mode.
+	//
+	// - FireWolf
+	// - 2019.06
+	//
+	
+	// Retrieve the framebuffer index
+	auto index = OSDynamicCast(OSNumber, framebuffer->getProperty("IOFBDependentIndex"))->unsigned32BitValue();
+	DBGLOG("igfx", "SC: GetDPCDInfo() DInfo: [FB%d] called with controller at 0x%llx and framebuffer at 0x%llx.", index, that, framebuffer);
+	
+	// Retrieve the user preference
+	LSPCON *lspcon = nullptr;
+	auto pmode = framebufferLSPCONGetPreferredMode(index);
+	
+	// Guard: Check whether this framebuffer connector has an onboard LSPCON chip
+	if (!framebufferHasLSPCON(index)) {
+		// No LSPCON chip associated with this connector
+		DBGLOG("igfx", "SC: GetDPCDInfo() DInfo: [FB%d] No LSPCON chip associated with this framebuffer.", index);
+		goto org;
+	}
+	
+	// Guard: Check whether the LSPCON driver has already been initialized for this framebuffer
+	if (framebufferHasLSPCONInitialized(index)) {
+		// Already initialized
+		lspcon = framebufferGetLSPCON(index);
+		DBGLOG("igfx", "SC: GetDPCDInfo() DInfo: [FB%d] LSPCON driver (at 0x%llx) has already been initialized for this framebuffer.", index, lspcon);
+		// Confirm that the adapter is running in preferred mode
+		if (lspcon->setModeIfNecessary(pmode) != kIOReturnSuccess) {
+			SYSLOG("igfx", "SC: GetDPCDInfo() Error: [FB%d] The adapter is not running in preferred mode. Failed to update the mode.", index);
+		}
+		// Wake up the native AUX channel if PCON mode is preferred
+		if (pmode == LSPCON::Mode::ProtocolConverter) {
+			if (lspcon->wakeUpNativeAUX() != kIOReturnSuccess)
+				SYSLOG("igfx", "SC: GetDPCDInfo() Error: [FB%d] Failed to wake up the native AUX channel.", index);
+		}
+		goto org;
+	}
+	
+	// User has specified the existence of onboard LSPCON
+	// Guard: Initialize the driver for this framebuffer
+	lspcon = new LSPCON(that, framebuffer, displayPath);
```

Comment:
Just as mentioned previously, most likely this wants to be LSPCon::create(that, framebuffer, displayPath), same with the delete below.